tests/subsys/net/lib/azure_iot_hub/src/main.c

```diff
@@ -0,0 +1,297 @@
+	zassert_true(strcmp(topic_data.prop_bag[0].value, "59765") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].key, "retry-after") == 0,
+			    NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].value, "3") == 0, NULL);
+}
+
+static void test_topic_parse_prop_bag_overload(void)
+{
+	int err;
+	const char *topic =
+		"$dps/registrations/res/204/?$rid=59765&retry-after=3&"
+		"test1=0x0&tes2&test3&test4&test5";
+	struct topic_parser_data topic_data = {
+		.topic = topic,
+		.topic_len = strlen(topic),
+		.type = TOPIC_TYPE_UNKNOWN,
+	};
+
+	err = azure_iot_hub_topic_parse(&topic_data);
+	zassert_equal(err, 0, NULL);
+	zassert_equal(topic_data.type, TOPIC_TYPE_DPS_REG_RESULT,
+		      "Incorrect topic type");
+	zassert_equal(topic_data.status, 204, NULL);
+	zassert_equal(topic_data.prop_bag_count, 5,
+		      "Incorrect property bag count");
+	zassert_true(strcmp(topic_data.prop_bag[0].key, "$rid") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[0].value, "59765") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].key, "retry-after") == 0,
+			    NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].value, "3") == 0, NULL);
+}
+
+static void test_topic_parse_unknown_topic(void)
+{
+	int err;
+	const char *topic =
+		"ive/never/seen/this/topic/before/?whos=this";
+	struct topic_parser_data topic_data = {
+		.topic = topic,
+		.topic_len = strlen(topic),
+		.type = TOPIC_TYPE_UNKNOWN,
+	};
+
+	err = azure_iot_hub_topic_parse(&topic_data);
+	zassert_equal(err, 0, NULL);
+	zassert_equal(topic_data.type, TOPIC_TYPE_UNKNOWN,
+		      "Incorrect topic type");
+	zassert_equal(topic_data.prop_bag_count, 0,
+		      "Incorrect property bag count");
+}
+
+static void test_topic_add_prop_bags(void)
+{
+	char *key1 = "key1";
+	char *val1 = "value1";
+	char *key2 = "key2";
+	char *val2 = "";
+	char *key3 = "key3";
+	struct azure_iot_hub_prop_bag bags[] = {
+		{
+			.key = key1,
+			.value = val1,
+		},
+		{
+			.key = key2,
+			.value = val2,
+		},
+		{
+			.key = key3,
+			.value = NULL,
+		},
+	};
+	char *prop_bag_str = azure_iot_hub_prop_bag_str_get(bags,
+							    ARRAY_SIZE(bags));
+
+	zassert_not_null(prop_bag_str, NULL);
+	zassert_equal(strcmp(prop_bag_str, "?key1=value1&key2=&key3"), 0,
+		      "Incorrect property bag string");
+	k_free(prop_bag_str);
+}
+
+static void test_topic_add_prop_bags_reverse(void)
+{
+	char *key1 = "key1";
+	char *value1 = "value1";
+	char *key2 = "key2";
+	char *value2 = "";
+	char *key3 = "key3";
+	struct azure_iot_hub_prop_bag bags[] = {
+		{
+			.key = key3,
+			.value = NULL,
+		},
+		{
+			.key = key2,
+			.value = value2,
+		},
+		{
+			.key = key1,
+			.value = value1,
+		},
+	};
+	char *prop_bag_str = azure_iot_hub_prop_bag_str_get(bags,
+							    ARRAY_SIZE(bags));
+
+	zassert_not_null(prop_bag_str, NULL);
+	zassert_equal(strcmp(prop_bag_str, "?key3&key2=&key1=value1"), 0,
+		      "Incorrect property bag string");
+	k_free(prop_bag_str);
```

Comment:
Maybe add a static inline function in ```azure_iot_hub_topic.h``` that implicitly frees ```property_bag_str``` type pointers? Makes it more intuitive looking at the API.

Reply:
Yep, that makes it more clear what happens. Added